### PR TITLE
new charm

### PR DESCRIPTION
**Pull request by kwmonroe**
This started as a quick switch from the `docker-registry` apt package to running a `registry:2` container under `docker.io`.

It ended up pretty much as a whole new charm.  It has tested well for me and is available in the store here:

https://jujucharms.com/u/containers/docker-registry/

### CODE REVIEW

File: reactive/docker_registry.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        hookenv.log('Sending {} and tls config to registry client.'.format(registry_url))
+        request.set_registry_config(registry_url=registry_url,
+                                    tls_enabled=tls_enabled,
+                                    tls_ca=tls_ca)
+    registry.mark_completed()
+
+
+@when('cert-provider.available')
+def request_certificates():
+    '''Request a new cert/key initally, and any time our SANs change.'''
```
Comment (ktsakalozos) [inline reactive/docker_registry.py]: typo

File: reactive/docker_registry.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    for request in registry.requests:
+        hookenv.log('Sending {} and tls config to registry client.'.format(registry_url))
+        request.set_registry_config(registry_url=registry_url,
+                                    tls_enabled=tls_enabled,
+                                    tls_ca=tls_ca)
+    registry.mark_completed()
+
+
+@when('cert-provider.available')
+def request_certificates():
```
Comment (ktsakalozos) [inline reactive/docker_registry.py]: This looks as if we "Request a new cert/key initally, and every 5 minutes check for a sans update".

File: reactive/docker_registry.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        cert_provider.request_server_cert(cert_cn, sans, cert_name)
+
+
+@when('charm.docker-registry.configured')
+@when('cert-provider.server.certs.changed')
+def write_certs():
+    '''Write cert data to our configured location.'''
+    cert_provider = endpoint_from_flag('cert-provider.server.certs.changed')
+    ca = cert_provider.root_ca_cert
+    cert = cert_provider.server_certs[0]  # only requested one
```
Comment (ktsakalozos) [inline reactive/docker_registry.py]: Is it possible to get the certificate is a non-0 position? https://github.com/juju-solutions/interface-tls-certificates/blob/master/requires.py#L174 What would happen if the same easyrsa is used by other deployments? Does this play any role?

Comment (ktsakalozos) [reply reactive/docker_registry.py]: Would a dictionary work better for what the interface returns?

File: lib/charms/layer/docker_registry.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+def configure_registry():
+    '''Recreate the docker registry config.yml.'''
+    charm_config = hookenv.config()
+    registry_config = {'version': '0.1'}
+    registry_config_file = '/etc/docker/registry/config.yml'
+
+    # Some files need to be volume mounted in the container. Keep track of
+    # those (recreate each time we configure). Regardless of the src location,
+    # we explictly mount them in the container under /etc/docker/registry.
```
Comment (ktsakalozos) [inline lib/charms/layer/docker_registry.py]: typo?

File: lib/charms/layer/docker_registry.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        cmd = ['docker', 'container', 'start', name]
+        try:
+            subprocess.check_call(cmd)
+        except subprocess.CalledProcessError:
+            hookenv.log('Could not start existing container: {}'.format(name),
+                        level=hookenv.ERROR)
+            raise
+    else:
+        # NB: config determines the port, but the container always listens to 5000 internally
+        cmd = ['docker', 'run', '-d', '-p', '{}:5000'.format(port), '--restart', 'unless-stopped']
```
Comment (ktsakalozos) [inline lib/charms/layer/docker_registry.py]: Do you think we should have done a systemd config?

File: reactive/docker_registry.py (hunk 2 above)
Comment (kwmonroe) [reply reactive/docker_registry.py]: This is true.  We allow people to configure http-host, which could affect the SANs list.  However, it's probably better to have a separate config handler for that and avoid running this every 5 minutes.  I'll investigate that approach.

File: reactive/docker_registry.py (hunk 3 above)
Comment (kwmonroe) [reply reactive/docker_registry.py]: Yup, we can use https://github.com/juju-solutions/interface-tls-certificates/blob/master/requires.py#L204 and look up the cert by CN.

File: lib/charms/layer/docker_registry.py (hunk 5 above)
Comment (kwmonroe) [reply lib/charms/layer/docker_registry.py]: I don't think so.  The `--restart` flag will ensure the container is started when the docker service comes up.

**Comment by ktsakalozos** [comment]
Lets merge it! LGTM +1

**Pull request merged**